### PR TITLE
Factoryze send mail

### PR DESCRIPTION
**Pull request by FanchTheSystem**
work in progress, but comment are welcome

Goal:
- start cleaning and factoryze and normalize mail sending process
- add some mail for project workflow

### CODE REVIEW

**Comment by osallou** [comment]
Will have further look but sounds good.

**Comment by FanchTheSystem** [comment]
So I have transfert the "user mail" fonction to utils.js, and use it in users.js and projects.js

There is also some user mail in tp.js and database.js and auth.js, I can propose the code change, but I will not be able to test it well. Do you want I do it ?

File: expire.js [hunk 1]
```diff
@@ -53,17 +53,15 @@ utils.init_db().then(async () => {
...
-                subject: CONFIG.general.name + ' account expiration: ' + user.uid, // Subject line
-                message: msg_activ, // plaintext body
-                html_message: msg_activ_html // html body
-            };
             try {
-                await notif.sendUser(mailOptions);
+                await utils.send_notif_mail({
+                    'name': 'expired',
+                    destinations: [CONFIG.general.support],
+                    subject: 'account expire: ' + user.uid
```
Comment (osallou) [inline expire.js]: please keep "expiration"

File: project_quota_expire.js [hunk 2]
```diff
@@ -38,40 +38,31 @@ utils.init_db().then(async () => {
...
-            try {
-                await notif.sendUser(mailOptions);
-            } catch(error) {
-                console.log(error);
-            }
-            mail_sent++;
-            if(mail_sent == notifs.length) {
-                process.exit(0);
-            }
+        let project = project[i];
```
Comment (osallou) [inline project_quota_expire.js]: replacing project by project[i] ? you mean project = projects[i]

File: project_quota_expire.js [hunk 3]
```diff
@@ -38,40 +38,31 @@ utils.init_db().then(async () => {
...
+            console.log(error);
         }
-        else {
-            console.log(notif);
-            mail_sent++;
-            if(mail_sent == notifs.length) {
-                process.exit(0);
-            }
+        mail_sent++;
+        if(mail_sent == notifs.length) {
```
Comment (osallou) [inline project_quota_expire.js]: the mail count and test below should not be needed anymore when we switched to async.
as we do an await on each mail sending, when loop is over, all mails are sent.

File: routes/users.js [hunk 4]
```diff
@@ -1021,14 +1007,15 @@ router.get('/user/:id/activate', async function(req, res) {
             return plugin_call(plugin_info, user.uid, user, session_user.uid);
             // eslint-disable-next-line no-unused-vars
         })).then(function(results){
-            return send_notif(mailOptions, fid, []);
+            res.send({msg: 'Activation in progress', fid: fid, error: []});
+            res.end();
+            return;
         }, function(err){
-            return send_notif(mailOptions, fid, err);
-        }).then(function(errs){
```
Comment (osallou) [inline routes/users.js]: mail is sent before plugin, in case of err, won't be notified of error

File: routes/users.js [hunk 5]
```diff
@@ -1662,18 +1648,16 @@ router.get('/user/:id/renew', async function(req, res){
                 return plugin_call(plugin_info, user.uid, user, session_user.uid);
                 // eslint-disable-next-line no-unused-vars
             })).then(function(results){
-                return send_notif(mailOptions, fid, []);
+                res.send({message: 'Activation in progress', fid: fid, error: []});
+                res.end();
+                return;
             }, function(err){
-                return send_notif(mailOptions, fid, err);
```
Comment (osallou) [inline routes/users.js]: same as previous, mail sent before plugin execution and possible failure

File: routes/utils.js [hunk 6]
```diff
@@ -430,3 +433,72 @@ function _freeId (objType, id) {
...
+
+
+// generic function for send notif (and email)
+
+function get_mail_config () {
+    var MAIL_CONFIG = {};
+    // todo: more and more ugly init...
+    if (CONFIG[MAILER]) { MAIL_CONFIG = CONFIG[MAILER]; }
+    // todo: find a cleaner way to allow registration if no mail are configured
+    if (!MAIL_CONFIG.origin) { MAIL_CONFIG.origin = 'nomail@nomail.org'; }
```
Comment (osallou) [inline routes/utils.js]: if no original email, means there is no contact for the tool/team. Can't think of any real use case where there is no contact email

File: routes/utils.js [hunk 7]
```diff
@@ -430,3 +433,72 @@ function _freeId (objType, id) {
...
+{
+    var MAIL_CONFIG = get_mail_config();
+    // todo: check if each option exist and use default value
+    let name = options['name'];
+    let destinations = options['destinations'];
+    let subject = GENERAL_CONFIG.name + ' ' + options['subject'];
+
+    //find message
+    let message = CONFIG.message[name].join('\n');
+    let html_message = CONFIG.message[name + '_html'].join('');
```
Comment (osallou) [inline routes/utils.js]: should we not manage case where name:
1) is not defined in config
2) we have *name* (text) but not *name*_html (html version). In this case use text only.

File: expire.js (hunk 1 above)
Comment (FanchTheSystem) [reply expire.js]: if I use "expiration", I will have to rename the current "expiration" as "expiration_warning" (there is 2 mail for expiration, one before and one after)

File: project_quota_expire.js (hunk 2 above)
Comment (FanchTheSystem) [reply project_quota_expire.js]: thank you, in fact it is notif, as notif array does not contain message anymore, now it contain project 
(most of time I try to minimize difference on code to keep readability of the change)

File: project_quota_expire.js (hunk 3 above)
Comment (FanchTheSystem) [reply project_quota_expire.js]: yes I think too, I will fix it

File: routes/users.js (hunk 4 above)
Comment (FanchTheSystem) [reply routes/users.js]: error was not sent by mail in previous version, like "fid", "err" was only pass as parameter to be used in the next ".then" (ie callback)

File: routes/utils.js (hunk 6 above)
Comment (FanchTheSystem) [reply routes/utils.js]: Yes , no real case, only with some specific test env, I will remove it

File: routes/utils.js (hunk 7 above)
Comment (FanchTheSystem) [reply routes/utils.js]: Yes I have add a comment with a  7 line before, so it is in my "todo list" ;)

File: expire.js (hunk 1 above)
Comment (osallou) [reply expire.js]: I meant "account expiration" or "account expired" for subject

File: routes/notif_nodemailer.js [hunk 8]
```diff
@@ -23,9 +23,9 @@ if(CONFIG.nodemailer){
 }
 
 
-function verify_transport () {
+async function verify_transport () {
     // eslint-disable-next-line no-unused-vars
-    transporter.verify(function(error, success) {
+    await transporter.verify(function(error, success) {
```
Comment (osallou) [inline routes/notif_nodemailer.js:28]: as there is an await here, there should be no callback function here, no? something like

    try {
        await transporter.verify()
        mail_set = true
        ....
    } catch() {

   }

and catch possible error , set mail_set etc... if success

File: routes/utils.js [hunk 9]
```diff
@@ -430,3 +434,91 @@ function _freeId (objType, id) {
...
+    let destinations = options['destinations'];
+    let subject = GENERAL_CONFIG.name + ' ' + options['subject'];
+
+    let message = 'Email message not found';
+
+    //find message
+    if (name && CONFIG.message[name]) {
+        message = CONFIG.message[name].join('\n');
+    } else {
+        logger.error('Email Message not found!');
```
Comment (osallou) [inline routes/utils.js:466]: should we not return from fonction here?

File: routes/users.js [hunk 10]
```diff
@@ -826,16 +809,18 @@ router.delete_user = async function(user, action_owner_id, message){
...
-        html_message: msg_del_html // html body
-    };
+    try {
+        await utils.send_notif_mail({
+            'name': 'deletion',
+            'destinations': msg_destinations,
+            'subject': 'account deletion: ' + user.uid
+        }, {
+            '#UID#': user.uid,
+            '#MSG#': mail_message
```
Comment (osallou) [inline routes/users.js:819]: there was before variable *action_owner_id* , you do not set it here (unless you changed template to remove it?)

File: routes/notif_nodemailer.js (hunk 8 above)
Comment (FanchTheSystem) [reply routes/notif_nodemailer.js]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/2dcd5caf9dac371d019970ebc5d1c7dc20e3c4cf

File: routes/users.js (hunk 10 above)
Comment (FanchTheSystem) [reply routes/users.js:820]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/e871307bafc84093fe6a8956a45d22e14e07ba0f

File: routes/utils.js (hunk 9 above)
Comment (FanchTheSystem) [reply routes/utils.js]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/85ec1bdc1ec57c5022281b1b1304d9c89ebd0b98

File: project_quota_expire.js (hunk 3 above)
Comment (FanchTheSystem) [reply project_quota_expire.js]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/c8b099ca292fbbd49df4c73e481dd9cfad49d827

File: expire.js (hunk 1 above)
Comment (FanchTheSystem) [reply expire.js]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/bca560ac99cd7eeb97dc32146e82c0fca13232d0

File: routes/utils.js (hunk 6 above)
Comment (FanchTheSystem) [reply routes/utils.js]: https://github.com/FanchTheSystem/genouestaccountmanager/commit/b208d9a245ff6d07270eff0f1430c17ebb2e185f

File: routes/utils.js (hunk 7 above)
Comment (FanchTheSystem) [reply routes/utils.js]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/85ec1bdc1ec57c5022281b1b1304d9c89ebd0b98

File: routes/users.js (hunk 5 above)
Comment (FanchTheSystem) [reply routes/users.js:1667]: now send mail and plugins are unlinked so they should not failed one because of the other

File: routes/users.js [hunk 11]
```diff
@@ -826,16 +809,19 @@ router.delete_user = async function(user, action_owner_id, message){
...
-    let mailOptions = {
-        origin: MAIL_CONFIG.origin, // sender address
-        destinations:  msg_destinations, // list of receivers
-        subject: GENERAL_CONFIG.name + ' account deletion: ' + user.uid, // Subject line
-        message: msg_del, // plaintext body
-        html_message: msg_del_html // html body
-    };
+    try {
+        await utils.send_notif_mail({
+            'name': 'deletion',
```
Comment (osallou) [inline routes/users.js:814]: would be better to name template (and according name in config) user_deletion, deletion is a bit too generic

File: test_expiration.js [hunk 12]
```diff
@@ -43,33 +43,28 @@ utils.init_db().then(async ()=>{
...
             } catch(error) {
                 console.log(error);
             }
             mail_sent++;
-            if(mail_sent == users.length) {
-                process.exit(0);
-            }
         }(i));
     }
     if(mail_sent == users.length) {
```
Comment (osallou) [inline test_expiration.js:67]: should add a else case with error log and exit(1)

File: routes/users.js [hunk 13]
```diff
@@ -1638,16 +1620,21 @@ router.get('/user/:id/renew', async function(req, res){
...
-            let mailOptions = {
-                origin: MAIL_CONFIG.origin, // sender address
-                destinations: [user.email], // list of receivers
-                subject: GENERAL_CONFIG.name + ' account reactivation', // Subject line
-                message: msg_activ, // plaintext body
-                html_message: msg_activ_html // html body
-            };
+
+            try {
+                utils.send_notif_mail({
```
Comment (osallou) [inline routes/users.js:1625]: missing await here I suppose

File: config/test.json [hunk 14]
```diff
@@ -166,6 +172,18 @@
...
       "<hr>"
     ],
+    "expired": [
+        "User #UID# has expired, updating account"
+    ],
+    "expired_html": [
+        "<p>User #UID# has expired, updating account</p>"
+    ],
+    "project_expiration": [
+        "Project quota for #NAME# will expire at #DATE#"
```
Comment (osallou) [inline config/test.json:182]: should just talk about project expiration, not quota, project may be more than a disk extension

File: config/test.json [hunk 15]
```diff
@@ -166,6 +172,18 @@
...
+    ],
+    "project_expiration_html": [
+        "<p>Project quota for #NAME# will expire at #DATE#</p>"
```
Comment (osallou) [inline config/test.json:185]: same as above

File: config/default.json.template [hunk 16]
```diff
@@ -243,6 +251,18 @@
...
       "<hr>"
     ],
+    "expired": [
+        "User #UID# has expired, updating account"
+    ],
+    "expired_html": [
+        "<p>User #UID# has expired, updating account</p>"
+    ],
+    "project_expiration": [
+        "Project quota for #NAME# will expire at #DATE#"
```
Comment (osallou) [inline config/default.json.template:261]: same as below comment (test.json) do not talk about quota

File: config/default.json.template [hunk 17]
```diff
@@ -243,6 +251,18 @@
...
+    ],
+    "project_expiration_html": [
+        "<p>Project quota for #NAME# will expire at #DATE#</p>"
```
Comment (osallou) [inline config/default.json.template:264]: same quota comment

File: config/test.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (FanchTheSystem) [reply config/test.json]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/aebfa84ebba26c26dc21888b0836aba0b3cd352a

Comment (FanchTheSystem) [reply config/test.json]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/aebfa84ebba26c26dc21888b0836aba0b3cd352a

File: config/default.json.template (hunk 16 above)
Comment (FanchTheSystem) [reply config/default.json.template]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/aebfa84ebba26c26dc21888b0836aba0b3cd352a

File: config/default.json.template (hunk 17 above)
Comment (FanchTheSystem) [reply config/default.json.template]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/aebfa84ebba26c26dc21888b0836aba0b3cd352a

File: routes/users.js (hunk 11 above)
Comment (FanchTheSystem) [reply routes/users.js]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/b2733a6e9e4b6a6e5e7945620b5c048218ff134c

File: config/test.json (hunk 15 above)
Comment (FanchTheSystem) [reply config/test.json]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/aebfa84ebba26c26dc21888b0836aba0b3cd352a

File: routes/users.js (hunk 13 above)
Comment (FanchTheSystem) [reply routes/users.js]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/4e49ab5ad621309ce7483b4b7de13d45644727e0

File: test_expiration.js [hunk 18]
```diff
@@ -43,33 +43,28 @@ utils.init_db().then(async ()=>{
...
         process.exit(0);
     }
```
[the remark this answers is not in the document]
Comment (FanchTheSystem) [reply test_expiration.js:72]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/038087cb327f6d9d3cbb3fa2247b8aaedf98ad7b

File: test_expiration.js (hunk 12 above)
Comment (FanchTheSystem) [reply test_expiration.js:67]: https://github.com/genouest/genouestaccountmanager/pull/166/commits/038087cb327f6d9d3cbb3fa2247b8aaedf98ad7b

**Comment by osallou** [comment]
I will be **really** busy from tomorrow till end of next week, will try to merge tomorrow but just in case, if there is no progress, ping me again

**Comment by FanchTheSystem** [comment]
> I will be **really** busy from tomorrow till end of next week, will try to merge tomorrow but just in case, if there is no progress, ping me again

great, please tag after merge

**Comment by osallou** [comment]
do not forget to rebuild the mycron docker image which depends on my image

**Pull request merged**